File: one-more-time/Lagrange/Lagrange.cs
using System;
using System.Collections.Generic;
using System.IO;

public class Point
{
    public double x, y;

    public Point(double x, double y)
    {
        this.x = x;
        this.y = y;
    }
}

public class Lagrange
{
    public static List<double> Zeros(int n)
    {
        List<double> arr = new List<double>();
        for (int i = 0; i < n; i++)
        {
            arr.Add(0);
        }
        return arr;
    }

    public static double Denom(int i, List<Point> points)
    {
        double p = 1;
        double x = points[i].x;
        for (int j = 0; j < points.Count; j++)
        {
            if (i != j)
            {
                p *= x - points[j].x;
            }
        }
        return p;
    }

    public static List<double> Interpolate(int i, List<Point> points)
    {
        List<double> coeff = Zeros(points.Count);
        coeff[0] = 1 / Denom(i, points);

        for (int k = 0; k < points.Count; k++)
        {
            if (k == i)
            {
                continue;
            }
            List<double> newCoeff = Zeros(points.Count);
            for (int j = (k < i) ? k + 1 : k; j >= 0; j--)
            {
                if (j < 2)
                {
                    newCoeff[j + 1] += coeff[j];
                    newCoeff[j] -= points[k].x * coeff[j];
                }
            }
            coeff = newCoeff;
        }
        return coeff;
    }

    public static List<double> Lagrange(List<Point> points)
    {
        List<double> polynom = Zeros(points.Count);
        List<double> coeff;

        for (int i = 0; i < points.Count; i++)
        {
            coeff = Interpolate(i, points);
            for (int k = 0; k < points.Count; k++)
            {
                polynom[points.Count - 1 - k] += points[i].y * coeff[k];
            }
        }
        return polynom;
    }

    public static void Main(string[] args)
    {
        List<Point> points = new List<Point>();
        using (StreamReader input = new StreamReader("ingput.txt"))
        {
            int degree = int.Parse(input.ReadLine());
            for (int i = 0; i < degree + 1; i++)
            {
                string[] info = input.ReadLine().Split();
                double x = double.Parse(info[0]);
                double y = double.Parse(info[1]);
                points.Add(new Point(x, y));
            }
        }

        List<double> final = Lagrange(points);
        using (StreamWriter output = new StreamWriter("otput.txt"))
        {
            foreach (double coeff in final)
            {
                output.Write(coeff + " ");
            }
        }

        Console.WriteLine("Lagrange coefficients successfully written.");
    }
}

File: one-more-time/Lagrange/Lagrange.py
#python 3.x
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def zeros(n):
    return [0] * n

def denom(i, points):
    p = 1
    x = points[i].x
    for j in range(len(points)):
        if i != j:
            p *= x - points[j].x
    return p

def interpolate(i, points):
    coeff = zeros(len(points))
    coeff[0] = 1 / denom(i, points)

    for k in range(len(points)):
        if k == i:
            continue
        new_coeff = zeros(len(points))
        for j in range(k + 1 if k < i else k, -1, -1):
            if j < 2:
                new_coeff[j + 1] += coeff[j]
                new_coeff[j] -= points[k].x * coeff[j]
        coeff = new_coeff
    return coeff

def lagrange(points):
    polynom = zeros(len(points))
    for i in range(len(points)):
        coeff = interpolate(i, points)
        for k in range(len(points)):
            polynom[len(points) - 1 - k] += points[i].y * coeff[k]
    return polynom

if __name__ == "__main__":
    points = []
    with open("ingput.txt", "r") as input:
        degree = int(input.readline())
        for _ in range(degree + 1):
            x, y = map(float, input.readline().split())
            points.append(Point(x, y))

    final_coeff = lagrange(points)
    with open("otput.txt", "w") as output:
        output.write(" ".join(map(str, final_coeff)))

    print("Lagrange coefficients successfully written.")
